luavm/src/lib/lua_values/mod.rs

```diff
@@ -0,0 +1,754 @@
+        if let LuaValKind::TABLE = this.kind() {
+            mark(unsafe { &*table_ptr(this.encoded_ptr) });
+        }
+    });
+}
+
+impl LuaVal {
+    /// Create an empty LuaVal which is equivalent to Nil.
+    pub fn new() -> LuaVal {
+        LuaVal {
+            encoded_ptr: 0,
+            owns: true,
+        }
+    }
+
+    /// Copies the pointer into a new LuaVal.
+    fn shallow_copy(&self) -> LuaVal {
+        LuaVal {
+            encoded_ptr: self.encoded_ptr,
+            owns: false,
+        }
+    }
+
+    /// Returns the type of the value store in the pointer.
+    fn kind(&self) -> LuaValKind {
+        LuaValKind::from(self.encoded_ptr)
+    }
+
+    /// Returns true if this value can be considered a float, or false otherwise.
+    pub fn is_float(&self) -> bool {
+        match self.kind() {
+            LuaValKind::FLOAT | LuaValKind::STR => true,
+            _ => false,
+        }
+    }
+
+    /// Attempts to convert this value to a float.
+    pub fn to_float(&self) -> Result<f64, LuaError> {
+        unsafe {
+            match self.kind() {
+                LuaValKind::INT => Ok(*(i64_ptr(self.encoded_ptr)) as f64),
+                LuaValKind::FLOAT => Ok(*(f64_ptr(self.encoded_ptr))),
+                LuaValKind::STR => (*(str_ptr(self.encoded_ptr)))
+                    .parse()
+                    .map_err(|_| LuaError::FloatConversionErr),
+                _ => Err(LuaError::FloatConversionErr),
+            }
+        }
+    }
+
+    /// Attempts to convert this value to an integer.
+    pub fn to_int(&self) -> Result<i64, LuaError> {
+        unsafe {
+            match self.kind() {
+                LuaValKind::INT => Ok(*(i64_ptr(self.encoded_ptr))),
+                _ => Err(LuaError::IntConversionErr),
+            }
+        }
+    }
+
+    /// Sets this to a new value.
+    pub fn set(&mut self, mut val: LuaVal) {
+        // exchange the pointers so that when `val` goes out of scope the old value
+        // is released and not leaked
+        let new_address = val.encoded_ptr;
+        val.encoded_ptr = self.encoded_ptr;
+        self.encoded_ptr = new_address;
+    }
+
+    /// Sets the given attribute to a given value.
+    pub fn set_attr(&mut self, attr: &str, val: LuaVal) -> Result<(), LuaError> {
+        if let LuaValKind::TABLE = self.kind() {
+            Ok(unsafe {
+                (*table_ptr(self.encoded_ptr)).set(attr, val);
+            })
+        } else {
+            Err(LuaError::SetAttrErr)
+        }
+    }
+
+    /// Gets the value of the given attribute.
+    pub fn get_attr(&self, attr: &str) -> Result<LuaVal, LuaError> {
+        if let LuaValKind::TABLE = self.kind() {
+            Ok(unsafe { (*table_ptr(self.encoded_ptr)).get_attr(attr) })
+        } else {
+            Err(LuaError::GetAttrErr)
+        }
+    }
+
+    pub fn add(&self, other: &LuaVal) -> Result<LuaVal, LuaError> {
+        Ok(if self.is_float() || other.is_float() {
```

Comment:
I'm not convinced by this: if I have a large 64 bit int and convert it to a float, it might not be representable. I fear you need to do something like:

```rust
match (self.kind(), other.kind) {
    (Int, Int) => ...,
    (Int, Float) => ...,
    (Float, Int) => ...,
    (Float, Float) => ...
}
```

the tricky cases, I think, are the middle two.

Reply:
I think it should be fine the way it is now, if I address the point below.